### PR TITLE
Reconnect to DB after timeout for Notify command. Fixes #14479

### LINKED ISSUE #14479: Notify command do not reconnect after DB connection has timed out
My MariaDB server has an 8 hour wait_timeout.

With "occ files_external:notify" running for 8 hours the process dies due to "MySQL error 2006: mysql server has gone away". Which means after 8 hours some change events are not registered.

We need to reconnect when that happens. PR is on its way :)
@icewind1991

### PR DESCRIPTION
**Pull request by ariselseng**
Fix #14479 

Every 8th hour I got a "2006 Mysql server has gone away". It was due to a 8 hour wait_time.
Since the Notify command should be running 24/7 increasing wait_time won't help much. Change events that triggers this never get registered and filecache will be wrong/out of sync.

This PR makes it reconnect in case of timeout. I borrowed the reconnect logic from: https://github.com/nextcloud/server/blob/1cfa870821c92bdd6854171c8c073ed16529b6df/apps/files/lib/Command/Scan.php#L300

I also needed to move updateQuery to a method due to it being a prepared statement from a dead connection and it needed to be remade after reconnect.

### CODE REVIEW

File: apps/files_external/lib/Command/Notify.php [hunk 1]
```diff
@@ -188,6 +193,39 @@ private function logUpdate(IChange $change, OutputInterface $output) {
...
+	}
+
+	/**
+	 * @return \OCP\IDBConnection
+	*/
+	private function reconnectToDatabase(IDBConnection $connection, OutputInterface $output) {
+		try {
+			$connection->close();
+		} catch (\Exception $ex) {
+			$output->writeln("<info>Error while disconnecting from database: {$ex->getMessage()}</info>");
```
Comment (MorrisJobke) [inline apps/files_external/lib/Command/Notify.php]: Please log this exception also. You need to inject the logger into the constructor of this class and then you can use it here with something like:

```php
$this->log->logException($ex, ['app' => 'files_external']);
```

File: apps/files_external/lib/Command/Notify.php [hunk 2]
```diff
@@ -188,6 +193,39 @@ private function logUpdate(IChange $change, OutputInterface $output) {
...
+		}
+		while (!$connection->isConnected()) {
+			try {
+				$connection->connect();
+			} catch (\Exception $ex) {
+				$output->writeln("<info>Error while re-connecting to database: {$ex->getMessage()}</info>");
```
Comment (MorrisJobke) [inline apps/files_external/lib/Command/Notify.php]: Same here.

File: apps/files_external/lib/Command/Notify.php (hunk 1 above)
Comment (MorrisJobke) [reply apps/files_external/lib/Command/Notify.php]: You can also specify a custom message then:

```php
$this->log->logException($ex, ['app' => 'files_external', 'message' => 'Error while disconnecting from DB', 'level' => 1]);
```

This logs with the message and as log level INFO.

Comment (cowai) [reply apps/files_external/lib/Command/Notify.php]: Got it!

Comment (MorrisJobke) [reply apps/files_external/lib/Command/Notify.php]: But maybe a warning is better suited here.

Comment (MorrisJobke) [reply apps/files_external/lib/Command/Notify.php]: Log levels are also available via constants for better readability: https://github.com/nextcloud/server/blob/38a90130ce425d531a804dff591df4a883de3154/lib/public/ILogger.php#L39-L58

File: apps/files_external/lib/Command/Notify.php (hunk 2 above)
Comment (cowai) [reply apps/files_external/lib/Command/Notify.php]: Yes. Thank you @MorrisJobke :+1:

**Comment by MorrisJobke** [comment]
Mind to add the sign-off message to your last commit: https://github.com/nextcloud/server/pull/14480/checks?check_run_id=72120851

**Comment by cowai** [comment]
@MorrisJobke woops. Done!

File: apps/files_external/lib/Command/Notify.php [hunk 3]
```diff
@@ -47,17 +50,15 @@ class Notify extends Base {
 	private $connection;
 	/** @var \OCP\DB\QueryBuilder\IQueryBuilder */
 	private $updateQuery;
+	/** @var ILogger */
+	private $log;
```
Comment (kesselb) [inline apps/files_external/lib/Command/Notify.php]: `$log` should be `$logger` because you use it that way (e.g. line 60, line 167)

**Review by kesselb — CHANGES_REQUESTED** [review]

File: apps/files_external/lib/Command/Notify.php (hunk 3 above)
Comment (cowai) [reply apps/files_external/lib/Command/Notify.php]: Thanks. Good catch!

**Review by icewind1991 — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Pull request merged**